File: auth-client/python/src/mcp_auth_client/token_exchange.py
import secrets
import time
from dataclasses import dataclass
from typing import Any
from urllib.parse import urlparse

import httpx
import jwt

from .cache import BoundedTokenCache


@dataclass(frozen=True, slots=True)
class TokenSet:
    access_token: str
    token_type: str
    expires_at: float
    audience: str
    scopes: frozenset[str]
    refresh_token: str | None = None

    @classmethod
    def from_response(cls, data: dict[str, Any], audience: str) -> "TokenSet":
        token = data.get("access_token")
        if not isinstance(token, str):
            raise ValueError("token response has no access_token")
        expires_in = data.get("expires_in", 300)
        if not isinstance(expires_in, (int, float)) or expires_in <= 0:
            raise ValueError("token response has invalid expires_in")
        return cls(
            token,
            str(data.get("token_type", "Bearer")),
            time.time() + float(expires_in),
            audience,
            frozenset(str(data.get("scope", "")).split()),
            data.get("refresh_token"),
        )


class PrivateKeyJWTClientAuth:
    def __init__(
        self, client_id: str, private_key_pem: str | bytes, key_id: str, audience: str | None = None
    ) -> None:
        self.client_id, self.private_key_pem, self.key_id = client_id, private_key_pem, key_id
        self.audience = audience

    def assertion(self, token_endpoint: str, lifetime: int = 300) -> str:
        now = int(time.time())
        return str(
            jwt.encode(
                {
                    "iss": self.client_id,
                    "sub": self.client_id,
                    "aud": self.audience or token_endpoint,
                    "iat": now,
                    "exp": now + lifetime,
                    "jti": secrets.token_urlsafe(16),
                },
                self.private_key_pem,
                algorithm="RS256",
                headers={"kid": self.key_id, "typ": "JWT"},
            )
        )

    def form_fields(self, token_endpoint: str) -> dict[str, str]:
        return {
            "client_id": self.client_id,
            "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            "client_assertion": self.assertion(token_endpoint),
        }


class TokenExchangeClient:
    def __init__(
        self,
        token_endpoint: str,
        client_auth: PrivateKeyJWTClientAuth | None = None,
        http_client: httpx.AsyncClient | None = None,
        cache: BoundedTokenCache | None = None,
    ) -> None:
        if urlparse(token_endpoint).scheme not in {"https", "http"}:
            raise ValueError("token endpoint must be an absolute URL")
        self.token_endpoint, self.client_auth = token_endpoint, client_auth
        self._http_client, self._owns_client = http_client, http_client is None
        self.cache = cache or BoundedTokenCache()

    async def __aenter__(self) -> "TokenExchangeClient":
        if self._http_client is None:
            self._http_client = httpx.AsyncClient(follow_redirects=False)
        return self

    async def __aexit__(self, *args: object) -> None:
        if self._owns_client and self._http_client is not None:
            await self._http_client.aclose()

    async def exchange(
        self,
        subject_token: str,
        audience: str,
        scopes: set[str] | frozenset[str] = frozenset(),
        use_cache: bool = True,
    ) -> TokenSet:
        key = f"{hash(subject_token)}|{audience}|{' '.join(sorted(scopes))}"
        if use_cache:
            cached = self.cache.get(key)
            if isinstance(cached, TokenSet):
                return cached
        if self._http_client is None:
            raise RuntimeError(
                "TokenExchangeClient must be used as an async context manager or "
                "given an httpx client"
            )
        form = {
            "grant_type": "urn:ietf:params:oauth:grant-type:token-exchange",
            "subject_token": subject_token,
            "subject_token_type": "urn:ietf:params:oauth:token-type:access_token",
            "requested_token_type": "urn:ietf:params:oauth:token-type:access_token",
            "audience": audience,
            "scope": " ".join(sorted(scopes)),
        }
        if self.client_auth:
            form.update(self.client_auth.form_fields(self.token_endpoint))
        response = await self._http_client.post(self.token_endpoint, data=form)
        response.raise_for_status()
        token = TokenSet.from_response(response.json(), audience)
        if use_cache:
            self.cache.put(key, token, token.expires_at)
        return token

File: auth-client/go/mcpauth/token_exchange.go
package mcpauth

import (
	"bytes"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

type PrivateKeyJWTClientAuth struct {
	ClientID   string
	PrivateKey *rsa.PrivateKey
	KeyID      string
}

func (a PrivateKeyJWTClientAuth) Assertion(endpoint string) (string, error) {
	now := time.Now().Unix()
	header := b64JSON(map[string]any{"typ": "JWT", "alg": "RS256", "kid": a.KeyID})
	payload := b64JSON(map[string]any{"iss": a.ClientID, "sub": a.ClientID, "aud": endpoint, "iat": now, "exp": now + 300, "jti": randomString()})
	message := header + "." + payload
	digest := sha256.Sum256([]byte(message))
	signature, err := rsa.SignPKCS1v15(rand.Reader, a.PrivateKey, crypto.SHA256, digest[:])
	if err != nil {
		return "", err
	}
	return message + "." + base64.RawURLEncoding.EncodeToString(signature), nil
}

type TokenExchangeClient struct {
	Endpoint   string
	HTTPClient *http.Client
	ClientAuth *PrivateKeyJWTClientAuth
	Cache      *TokenCache
	// AllowInsecure permits a non-https token endpoint. Leave it false so
	// subject tokens and client assertions are not posted over cleartext.
	AllowInsecure bool
}

var defaultExchangeClient = &http.Client{Timeout: 10 * time.Second, CheckRedirect: refuseRedirect}

func refuseRedirect(*http.Request, []*http.Request) error {
	return errors.New("token endpoint must not redirect")
}

func exchangeCacheKey(subjectToken, audience string, scopes []string) string {
	sorted := append([]string(nil), scopes...)
	sort.Strings(sorted)
	sum := sha256.Sum256([]byte(subjectToken + "\x00" + audience + "\x00" + strings.Join(sorted, " ")))
	return hex.EncodeToString(sum[:])
}

func (c *TokenExchangeClient) Exchange(subjectToken, audience string, scopes []string) (CachedToken, error) {
	parsed, err := url.Parse(c.Endpoint)
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return CachedToken{}, fmt.Errorf("token endpoint must be an absolute URL")
	}
	if parsed.Scheme != "https" && !c.AllowInsecure {
		return CachedToken{}, fmt.Errorf("token endpoint must use https")
	}
	key := exchangeCacheKey(subjectToken, audience, scopes)
	if c.Cache != nil {
		if token, ok := c.Cache.Get(key); ok {
			return token, nil
		}
	}
	form := url.Values{"grant_type": {"urn:ietf:params:oauth:grant-type:token-exchange"}, "subject_token": {subjectToken}, "subject_token_type": {"urn:ietf:params:oauth:token-type:access_token"}, "requested_token_type": {"urn:ietf:params:oauth:token-type:access_token"}, "audience": {audience}, "scope": {strings.Join(scopes, " ")}}
	if c.ClientAuth != nil {
		assertion, err := c.ClientAuth.Assertion(c.Endpoint)
		if err != nil {
			return CachedToken{}, err
		}
		form.Set("client_id", c.ClientAuth.ClientID)
		form.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
		form.Set("client_assertion", assertion)
	}
	client := c.HTTPClient
	if client == nil {
		client = defaultExchangeClient
	}
	request, err := http.NewRequest(http.MethodPost, c.Endpoint, bytes.NewBufferString(form.Encode()))
	if err != nil {
		return CachedToken{}, err
	}
	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	// 307 and 308 replay the body, so a redirect from an https endpoint to an
	// http one would post the subject token and client assertion in cleartext.
	// Refuse every redirect: a token endpoint has no reason to move. A caller
	// that set its own policy keeps it.
	if client.CheckRedirect == nil {
		copied := *client
		copied.CheckRedirect = refuseRedirect
		client = &copied
	}
	response, err := client.Do(request)
	if err != nil {
		return CachedToken{}, err
	}
	defer response.Body.Close()
	if response.StatusCode/100 != 2 {
		return CachedToken{}, fmt.Errorf("token exchange returned %d", response.StatusCode)
	}
	var result struct {
		AccessToken string `json:"access_token"`
		TokenType   string `json:"token_type"`
		ExpiresIn   int    `json:"expires_in"`
		Scope       string `json:"scope"`
	}
	if err := json.NewDecoder(response.Body).Decode(&result); err != nil || result.AccessToken == "" {
		return CachedToken{}, fmt.Errorf("invalid token exchange response")
	}
	token := CachedToken{AccessToken: result.AccessToken, TokenType: result.TokenType, Audience: audience, ExpiresAt: time.Now().Add(time.Duration(result.ExpiresIn) * time.Second), Scopes: strings.Fields(result.Scope)}
	if c.Cache != nil {
		c.Cache.Put(key, token)
	}
	return token, nil
}
func b64JSON(value map[string]any) string {
	encoded, _ := json.Marshal(value)
	return base64.RawURLEncoding.EncodeToString(encoded)
}
func randomString() string {
	var value [16]byte
	_, _ = rand.Read(value[:])
	return base64.RawURLEncoding.EncodeToString(value[:])
}
